### DISCUSSION TITLE
img2img upscale by tiles

### DISCUSSION THREAD

**Discussion by playlogitech** [comfyanonymous/ComfyUI#333] [Q&A]
is it possible in comfy? Every pro workflow rn is about details boost trough tile upscale

**Comment by WASasquatch** [comment]
Wouldn't you just upscale anyway and use tiled VAE?

**Reply by throwaway-mezzo-mix** [reply]
I think he probably means the thing you can do in automatic1111 where the actual generation is done in tiles. (i know the built-in img2img SD upscale script does this, probably some more things too)
tiled vae is great but it doesn't help much if keeping the desired latent resolution in memory is too much and causes an OOM.

**Reply by playlogitech** [reply]
https://2ch.hk/ai/src/191780/16801925155242.png

current tile upscale looks like a small factory in factorio game and this is just for 4 tiles, you can only imagine how it gonna look like with more tiles, possible but makes no sense

**Reply by WASasquatch** [reply]
Ah, you mean the GO BIG method I added to Easy Diffusion from ProgRockDiffusion. That would indeed be handy. I did find the results were hard to control, though I never tried A1111 version. It probably uses img2img color fix so they don't vary.

**Comment by GrimblyGorn** [comment]
Ya this is something I would like as well. I've tried a few similar setups to the one depicted and they seem very daunting to contemplate expanding. More so when considering how large I would need them to expand to. It does seem like something that could be built into 1 or 2 nodes that could work together and achieve the same results.

Along the same lines, an infinity outpaint node would be nice. Something that adds a pad buffer and grows the images but only processes those new chunks and does so on each individually, then it loops and grows the image again. Then you can give it the inputs and set the growth rate (padding size), set the desired final size for the image, and it can just sit there looping and expanding the image. Having it only process the new chunks individually should allow for growing these images as large as you want with no vram issues.

**Reply by WASasquatch** [reply]
Not sure an infinite canvas would be possible. The system is linear to a end point. So you'd have to manually set all that up to a final result, every run. Otherwise I'd image you'd need a canvas node, and then a layer node that layers the new result on the open canvas, meaning you'd have to still manually set up the location the result is applied to. And there is no backup, it's linear in one direction, so you'd need a node to slice up the canvas and feed into img2img to generate a new tile to then again, layer on to the canvas, but not sure if that is even possible without recursive issues

**Comment by playlogitech** [comment]
![16803585386870](https://user-images.githubusercontent.com/121442734/229295040-59838b51-c89e-4ad4-bd0a-d5f8680b4511.png)
![16803585386911](https://user-images.githubusercontent.com/121442734/229295051-16376622-5083-4a4e-8878-552bbe3ea0ea.png)
![16803585386972](https://user-images.githubusercontent.com/121442734/229295061-4a4fbc84-5a4e-4222-8fcb-86f6a1867aef.png)
![16803585387013](https://user-images.githubusercontent.com/121442734/229295064-f91ea8dc-d14e-408e-afc7-4892e597eaa4.png)

well i really tried, but my head is about to explode already, because on last stage i get high quality drop, probably due to attempt of too low resolution upscale, i simply cant go higher on my 4090, because there is always OOM error, comfy do not release the memory from previously made steps. Maybe you guys can have more success than me. 

Here is my workflow:

https://drive.google.com/file/d/1fv77SX03mzUV1iID2xDjwHrHZ4IF4joz/view?usp=sharing

by any chance can you dig into it? @comfyanonymous

**Reply by WASasquatch** [reply]
It should release the memory. The nodes are executed and then their memory cleared. Least that's how custom nodes run. Every run of Comfy it's like loading up the nodes for the first time, cause none of my classes can hold on to any data for next run.

**Reply by SLAPaper** [reply]
wow this workflow is so scary
I think it should pack into a single node for simplicity

**Reply by Ferniclestix** [reply]
<img width="1549" alt="multiupscaler3" src="https://github.com/comfyanonymous/ComfyUI/assets/128430314/3f81c18d-1027-4324-a84a-de9c96b1add6">
do it all in latent space, its less complex and will save massively on memory usage. I keep mine laid out in the same position as each segment so its easier to build as each section is laid out along the structure of the final image. but yeah, would be so much easier in a single node lol.

**Reply by WASasquatch** [reply]
Using latent space may cause artifacts though. Latent scaling is very lossy it seems. If the image tensors weren't on GPU this wouldn't be much an issue, as the images would be in RAM not hogging up the VRAM.

**Reply by Ferniclestix** [reply]
the trick is to add a sampler after a latent upscale,  it is a lot more like trying to control an exited ferret than standard upscaling but you can get very high res images this way on lower spec hardware.

**Reply by WASasquatch** [reply]
> the trick is to add a sampler after a latent upscale, it is a lot more like trying to control an exited ferret than standard upscaling but you can get very high res images this way on lower spec hardware.

Yeah the VAE encoding process is a beast. Seems very wasteful on resources.

**Reply by BlenderNeko** [reply]
for those that missed it. [tiled Ksamplers](https://github.com/BlenderNeko/ComfyUI_TiledKSampler) have been around for a little while now.

**Reply by Ferniclestix** [reply]
Hah, no Im blind, thanks a bunch for making it, ill integrate it into my planz now >:)
